# Notebook de Análise
Bem-vindo!

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn import datasets

In [2]:
url = "/Users/Ccesconetto/Downloads/Analise1/data/raw/vendas_ecommerce_sujo.csv"
data = pd.read_csv(url)
df = pd.DataFrame(data)
df

,id_pedido,data_pedido,categoria,estado,quantidade,preco_unitario,metodo_pagamento,status_pedido,avaliacao
0,1207,2024-05-17,Livros,BA,9.0,110.25,PIX,Em trânsito,NaN
1,1034,2023-03-22,Alimentos,SC,2.0,612.81,Boleto,Entregue,5.0
2,1164,02-02-24,Alimentos,RS,5.0,"R$ 507,87",Cartão de Débito,Pendente,1.0
3,1079,10/07/2023,Alimentos,RS,2.0,434.87,PIX,Cancelado,NaN
4,1094,2023-08-16,Roupas,BA,1.0,185.32,Cartão de Débito,Pendente,4.0
...,...,...,...,...,...,...,...,...,...
310,1189,03-04-24,Roupas,SC,9.0,394.87,PIX,Cancelado,5.0
311,1072,2023-06-23,Eletrônicos,PR,4.0,43.44,PIX,Entregue,4.0
312,1107,2023-09-16,Livros,RS,2.0,739.04,NaN,Entregue,4.0
313,1271,2024-10-21,ROUPAS,GO,4.0,327.7,Boleto,Entregue,5.0


In [3]:
df.info()
df.describe()
df.dtypes
df.columns

<class 'pandas.DataFrame'>
RangeIndex: 315 entries, 0 to 314
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id_pedido         315 non-null    int64  
 1   data_pedido       315 non-null    str    
 2   categoria         315 non-null    str    
 3   estado            299 non-null    str    
 4   quantidade        306 non-null    float64
 5   preco_unitario    315 non-null    str    
 6   metodo_pagamento  303 non-null    str    
 7   status_pedido     305 non-null    str    
 8   avaliacao         271 non-null    float64
dtypes: float64(2), int64(1), str(6)
memory usage: 22.3 KB


Index(['id_pedido', 'data_pedido', 'categoria', 'estado', 'quantidade',
       'preco_unitario', 'metodo_pagamento', 'status_pedido', 'avaliacao'],
      dtype='str')

In [4]:
df["data_pedido"]
df["data_pedido"].isnull().sum()
# 95 NaN values before the cleaning process;

df["data_pedido_clean"] = pd.to_datetime(
    df["data_pedido"], 
    format="mixed",
    dayfirst=True,
    errors="coerce")

df["data_pedido_clean"].dtypes
df["data_pedido_clean"].isnull().sum()
# 0 NaN values after the cleaning process;

np.int64(0)

# Comandos:
- .isnull().sum() → Conta o total de valores nulos em cada coluna
- pd.to_datetime() → Converte string → datetime64
- format="mixed" → Detecta múltiplos formatos de data automaticamente
- errors="coerce" → Transforma valores inválidos em NaN (ou NaT para datas)

In [ ]:
df["preco_unitario"]
df["preco_unitario"].isnull().sum()
# 0 NaN values before the cleaning process;

def normalizar_preco(valor: str) -> str:
    valor = str(valor).strip().replace("R$", "").strip()  
    
    if "," in valor:
        valor = valor.replace(".", "").replace(",", ".")  
    
    return valor
    return valor    

df["preco_unitario_clean"] = (
    df["preco_unitario"]
    .apply(normalizar_preco)
    .pipe(pd.to_numeric, errors="coerce")
    .astype(float)
)

df["preco_unitario_clean"].isnull().sum()  
# 0 NaN values after the cleaning process;


np.int64(0)

# Função normalizar_preco

## Definição
- `def normalizar_preco(valor: str) -> str` → define a função, recebe e retorna `str`

## Limpeza inicial
- `str(valor)` → converte para string, protege contra `float` e `NaN`
- `.strip()` → remove espaços nas bordas
- `.replace("R$", "")` → remove o prefixo de moeda
- `.strip()` → remove o espaço que sobrou após o `R$`

## Conversão de formato
- `if "," in valor` → identifica formato pt-BR (`"507,87"`) vs neutro (`"110.25"`)
- `.replace(".", "")` → remove separador de milhar: `"1.235,50"` → `"1235,50"`
- `.replace(",", ".")` → troca vírgula decimal por ponto: `"1235,50"` → `"1235.50"`
- `return valor` → devolve a string normalizada

## Pipeline
- `.apply(normalizar_preco)` → executa a função linha por linha na coluna
- `.pipe(pd.to_numeric, errors="coerce")` → converte string → número; inválidos → NaN
- `.astype(float)` → garante tipo final `float64`

In [6]:
df["categoria"].dtypes
df["categoria"].unique()
df["categoria"].value_counts()
df["categoria_clean"] = df["categoria"].str.title()
df["categoria_clean"].unique()
df

,id_pedido,data_pedido,categoria,estado,quantidade,preco_unitario,metodo_pagamento,status_pedido,avaliacao,data_pedido_clean,preco_unitario_clean,categoria_clean
0,1207,2024-05-17,Livros,BA,9.0,110.25,PIX,Em trânsito,NaN,2024-05-17,110.25,Livros
1,1034,2023-03-22,Alimentos,SC,2.0,612.81,Boleto,Entregue,5.0,2023-03-22,612.81,Alimentos
2,1164,02-02-24,Alimentos,RS,5.0,"R$ 507,87",Cartão de Débito,Pendente,1.0,2024-02-02,NaN,Alimentos
3,1079,10/07/2023,Alimentos,RS,2.0,434.87,PIX,Cancelado,NaN,2023-07-10,434.87,Alimentos
4,1094,2023-08-16,Roupas,BA,1.0,185.32,Cartão de Débito,Pendente,4.0,2023-08-16,185.32,Roupas
...,...,...,...,...,...,...,...,...,...,...,...,...
310,1189,03-04-24,Roupas,SC,9.0,394.87,PIX,Cancelado,5.0,2024-04-03,394.87,Roupas
311,1072,2023-06-23,Eletrônicos,PR,4.0,43.44,PIX,Entregue,4.0,2023-06-23,43.44,Eletrônicos
312,1107,2023-09-16,Livros,RS,2.0,739.04,NaN,Entregue,4.0,2023-09-16,739.04,Livros
313,1271,2024-10-21,ROUPAS,GO,4.0,327.7,Boleto,Entregue,5.0,2024-10-21,327.70,Roupas


# Comandos:
- df[""].unique() → Retorna todos os valores únicos da coluna
- df[""].value_counts() → Contagem de ocorrências de cada valor único
- df[""].str.title() → Primeira letra maiúscula de cada palavra do valor